lib/puppet/type/exec.rb

```diff
@@ -204,7 +204,7 @@ def value=(*values)
       validate do |user|
         if Puppet.features.microsoft_windows?
           self.fail "Unable to execute commands as other users on Windows"
-        elsif !Puppet.features.root? && resource.current_username() != user
+        elsif Facter.value(:id) != 'root' && Facter.value(:id) != user
```

Comment:
@joshcooper to summarize, rspec-puppet's `it { should compile }` does not work as regular user when using `exec { 'foo': user => 'bar' }` in your manifests without using fakeroot.